### ISSUE TITLE
index() uses sv_pos_b2u/sv_pos_u2b

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#10171]
Migrated from [rt.perl.org#72782](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=72782) (status was 'resolved')

Searchable as RT72782$

**Comment by p5pRT** [comment]
### From @nwc10

pp\.c
3412&#8203;:       sv\_pos\_u2b\(big\, &offset\, 0\);
3429&#8203;:       sv\_pos\_b2u\(big\, &retval\);

Both in pp\_index\.

This means that index will fail for UTF\-8 strings over 2GB\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @nwc10

pp\.c
3412&#8203;:       sv\_pos\_u2b\(big\, &offset\, 0\);
3429&#8203;:       sv\_pos\_b2u\(big\, &retval\);

Both in pp\_index\.

This means that index will fail for UTF\-8 strings over 2GB\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @cpansprout

Exact copy of 72782\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'new' to 'resolved'

**Comment by p5pRT** [comment]
### From @jkeenan

On Sat Feb 13 08&#8203;:15&#8203;:03 2010\, nicholas wrote&#8203;:
> pp\.c
> 3412&#8203;:       sv\_pos\_u2b\(big\, &offset\, 0\);
> 3429&#8203;:       sv\_pos\_b2u\(big\, &retval\);
> 
> Both in pp\_index\.
> 
> This means that index will fail for UTF\-8 strings over 2GB\.
> 
> Nicholas Clark
> 

I could not find any instance of 'sv\_pos\_u2b' or 'sv\_pos\_b2u' in the part of pp\.c dealing with pp\_index \(ll\. 3367\-3481\)\.

I suspected that might mean that the problem reported has been fixed\.  To test that hypothesis\, I created a branch \(smoke\-me/jkeenan/72782\-sv\_pos\_u2b\) in which I added the attached patch\.

Two points&#8203;:

1\. Is the patch a valid test of the problem?  If so\, does that mean the problem has been fixed?

2\. On my local machine\, running the patch slowed my machine to a crawl\, leading me to have to re\-start&#8203;:

real    0m29\.616s
user    0m0\.755s
sys    0m1\.273s

This is perhaps not surprising\, as the tests require the composition of two 2GB\-length strings\.

On dromedary\, the patch ran much more quickly&#8203;:

real    0m1\.598s
user    0m0\.827s
sys    0m0\.774s

So\, even if the tests are valid indicators of a solution\, we would probably not want to include these tests in a normal 'make test' run unless we knew they were not going to cripple the host machine\.

Suggestions?

Thank you very much\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>72782-0001-Add-some-tests-for-RT-72782.patch</summary>

```diff
From 20cbf152deafba0d0a87e30a7121c75936a8ebc2 Mon Sep 17 00:00:00 2001
From: James E Keenan <jkeenan@cpan.org>
Date: Sat, 21 Mar 2015 19:24:28 -0400
Subject: [PATCH] Add some tests for RT #72782.

The functions cited by Nicholas in this report seem not to be present any
more.  Add some tests which may indicate that the problem has been solved.
---
 t/op/index.t | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/t/op/index.t b/t/op/index.t
index 243cc1b..80d3912 100644
--- a/t/op/index.t
+++ b/t/op/index.t
@@ -8,7 +8,7 @@ BEGIN {
 }
 
 use strict;
-plan( tests => 122 );
+plan( tests => 124 );
 
 run_tests() unless caller;
 
@@ -257,3 +257,11 @@ is index('the main road', __PACKAGE__), 4,
 utf8::upgrade my $substr = "\x{a3}a";
 
 is index($substr, 'a'), 1, 'index reply reflects characters not octets';
+
+# RT #72782
+my $twogb = 1024**3;
+my $at2gb = "\x{a3}" . 'x' x ($twogb - 1);
+is index($at2gb, 'xxx'), 1, "Got index for 2GB length string";
+
+my $over2gb = "\x{a3}" . 'x' x ($twogb);
+is index($over2gb, 'xxx'), 1, "Got index for > 2GB length string";
-- 
1.9.1


```
</details>

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ilmari

"James E Keenan via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:

> On Sat Feb 13 08&#8203;:15&#8203;:03 2010\, nicholas wrote&#8203;:
>> pp\.c
>> 3412&#8203;:       sv\_pos\_u2b\(big\, &offset\, 0\);
>> 3429&#8203;:       sv\_pos\_b2u\(big\, &retval\);
>> 
>> Both in pp\_index\.
>> 
>> This means that index will fail for UTF\-8 strings over 2GB\.
>> 
>> Nicholas Clark
>> 
>
> I could not find any instance of 'sv\_pos\_u2b' or 'sv\_pos\_b2u' in the part of pp\.c dealing with pp\_index \(ll\. 3367\-3481\)\.
>
> I suspected that might mean that the problem reported has been fixed\.  To test that hypothesis\, I created a branch \(smoke\-me/jkeenan/72782\-sv\_pos\_u2b\) in which I added the attached patch\.

It was fixed by this commit&#8203;:

commit b464e2b7c8addfdd9b22b9a5949a89db7c73e43c
Author&#8203;: Tony Cook \<tony@&#8203;develop\-help\.com>
Date&#8203;:   2014\-04\-15 03&#8203;:57&#8203;:57 \+0200

&nbsp;    fix the I32 bug for index\(\) and rindex\(\)


> So\, even if the tests are valid indicators of a solution\, we would
> probably not want to include these tests in a normal 'make test' run
> unless we knew they were not going to cripple the host machine\.

We already have this\, in the form of the PERL\_TEST\_MEMORY environment
variable\, which indicates how many gigabytes of memory tests are allowed
to use\. The above commit added t/bigmem/index\.t\, runs if
PERL\_TEST\_MEMORY >= 2\.

\-\- 
\- Twitter seems more influential \[than blogs\] in the 'gets reported in
&nbsp;  the mainstream press' sense at least\.               \- Matt McLeod
\- That'd be because the content of a tweet is easier to condense down
&nbsp;  to a mainstream media article\.                      \- Calle Dybedahl

**Comment by p5pRT** [comment]
### From @jkeenan

On Sun Mar 22 04&#8203;:41&#8203;:10 2015\, ilmari wrote&#8203;:
> "James E Keenan via RT" \<perlbug\-followup@&#8203;perl\.org> writes&#8203;:
> 
> > On Sat Feb 13 08&#8203;:15&#8203;:03 2010\, nicholas wrote&#8203;:
> >> pp\.c
> >> 3412&#8203;:       sv\_pos\_u2b\(big\, &offset\, 0\);
> >> 3429&#8203;:       sv\_pos\_b2u\(big\, &retval\);
> >>
> >> Both in pp\_index\.
> >>
> >> This means that index will fail for UTF\-8 strings over 2GB\.
> >>
> >> Nicholas Clark
> >>
> >
> > I could not find any instance of 'sv\_pos\_u2b' or 'sv\_pos\_b2u' in the
> > part of pp\.c dealing with pp\_index \(ll\. 3367\-3481\)\.
> >
> > I suspected that might mean that the problem reported has been fixed\.
> > To test that hypothesis\, I created a branch \(smoke\-me/jkeenan/72782\-
> > sv\_pos\_u2b\) in which I added the attached patch\.
> 
> It was fixed by this commit&#8203;:
> 
> commit b464e2b7c8addfdd9b22b9a5949a89db7c73e43c
> Author&#8203;: Tony Cook \<tony@&#8203;develop\-help\.com>
> Date&#8203;:   2014\-04\-15 03&#8203;:57&#8203;:57 \+0200
> 
> fix the I32 bug for index\(\) and rindex\(\)
> 
> 
> > So\, even if the tests are valid indicators of a solution\, we would
> > probably not want to include these tests in a normal 'make test' run
> > unless we knew they were not going to cripple the host machine\.
> 
> We already have this\, in the form of the PERL\_TEST\_MEMORY environment
> variable\, which indicates how many gigabytes of memory tests are
> allowed
> to use\. The above commit added t/bigmem/index\.t\, runs if
> PERL\_TEST\_MEMORY >= 2\.

Thanks for the feedback\.  On that basis\, I'm marking this ticket Resolved\.

\-\- 
James E Keenan \(jkeenan@&#8203;cpan\.org\)

**Comment by p5pRT** [comment]
@jkeenan - Status changed from 'open' to 'resolved'
